src/ssa/malloc_ssa.cpp

```diff
@@ -516,3 +518,199 @@ void allow_record_memleak(goto_modelt &goto_model)
              name.find("::record_may_leak")!=std::string::npos;
     });
 }
+
+/// Replaces usages of a symbol with another symbol in the given expression.
+void replace_symbol(exprt &expr, const irep_idt &old, const irep_idt &updated)
+{
+  if(expr.id()==ID_symbol)
+  {
+    auto &symbol_expr=to_symbol_expr(expr);
+    if(symbol_expr.get_identifier()==old)
+      symbol_expr.set_identifier(updated);
+  }
+  else
+    for(auto &op : expr.operands())
+      replace_symbol(op, old, updated);
+}
+
+/// Removes assignments to __CPROVER_memory_leak inside free function.
```

Comment:
I understand the core idea behind tracking leaks, however, 2LS currently doesn't use the non-deterministic choice, see https://github.com/diffblue/2ls/blob/master/src/ssa/malloc_ssa.cpp#L500 . It always sets `recod_may_leak` to true rather than it being a nondet bool. @viktormalik mentioned that this was introduced because some examples didn't work correctly with the nondet choice at the time of experimenting with memory safety analysis (and that it may be due to abstract objects?). It may be worth revisiting this and checking if it possibly isn't fixed now after the CBMC rebase, though... Also tracking objects individually could help with this.

This caused issues with unwinding, an example of this is `regression/memsafety/nondet_free_leak_kind` which I added. Without my workaround, 2LS would claim no memory leak, even though there is a leak in one of the possible paths because the `__CPROVER_memory_leak` variable is overridden in each allocation site.

Reply:
Marking as resolved for now, I'd suggest using my workaround and as a part of further experiments lets check if we cannot remove these hacks (perhaps the nondet tracking works correctly after the CBMC update?)

Reply:
I agree. Just for reference, as Frantisek says, I had to remove the nondet choice b/c for some reason, it didn't work in 2LS. I have a suspicion that it has something to do with dynamic objects, but would have to give it more thought.